zkml/src/lookup/context.rs

```diff
@@ -0,0 +1,355 @@
+//! File containg code for lookup witness generation.
+
+use std::collections::{BTreeSet, HashMap};
+
+use ff::Field;
+use ff_ext::ExtensionField;
+use serde::{Deserialize, Serialize, de::DeserializeOwned};
+use transcript::Transcript;
+
+use crate::{
+    Element,
+    activation::Relu,
+    commit::precommit::Context,
+    iop::ChallengeStorage,
+    lookup::logup_gkr::structs::LogUpInput,
+    model::{InferenceTrace, Layer},
+    quantization::{self, Fieldizer},
+};
+
+use super::logup_gkr::error::LogUpError;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+pub enum TableType {
+    Relu,
+    Range,
+}
+
+impl TableType {
+    fn get_merged_table_column<E: ExtensionField>(
+        &self,
+        column_separator: Element,
+    ) -> (Vec<Element>, Vec<Vec<E::BaseField>>) {
+        match self {
+            TableType::Relu => {
+                let (comb, field): (Vec<Element>, Vec<(E::BaseField, E::BaseField)>) =
+                    (*quantization::MIN..=*quantization::MAX)
+                        .map(|i| {
+                            let out = Relu::apply(i);
+                            let i_field: E = i.to_field();
+                            let out_field: E = out.to_field();
+                            (
+                                i + out * column_separator,
+                                (i_field.as_bases()[0], out_field.as_bases()[0]),
+                            )
+                        })
+                        .unzip();
+                let (col_one, col_two): (Vec<E::BaseField>, Vec<E::BaseField>) =
+                    field.into_iter().unzip();
+                (comb, vec![col_one, col_two])
+            }
+            TableType::Range => {
+                let (element_out, field): (Vec<Element>, Vec<E::BaseField>) = (0..1
+                    << *quantization::BIT_LEN)
+                    .map(|i| {
+                        let i_field: E = i.to_field();
+                        (i, i_field.as_bases()[0])
+                    })
+                    .unzip();
+                (element_out, vec![field])
```

Comment:
Couldn't we instead delegate the range and relu table evaluation to their own types and then do the magic for proving it with lookups here ? i.e. getting rid of `TableType` and to manually create the values. Because now for example we have RELU logic in two places.